xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_forward.h

```diff
@@ -569,6 +637,42 @@ struct AttentionKernel {
               (tb_tile_offset.n() * MM0::Mma::WarpCount::kN) +
                   (my_warp_id / MM0::Mma::WarpCount::kM)};
 
+      // multiply by scaling factor
+      accum = cutlass::multiplies<typename MM0::Mma::FragmentC>()(
+          p.scale, accum);
+
+      // apply attention bias if applicable
+      if (p.attn_bias_ptr != nullptr) {
+        // load bias tile Bij into shared memory
+        typename MM0::BiasLoader::GmemTileIterator bias_iter(
+          {cutlass::layout::RowMajor(p.bias_strideM)},
+          // attn_bias_pointer points to matrix of size (n_queries, n_keys)
+          // for the relevant batch_id and head_id
+          p.attn_bias_ptr + query_start * p.bias_strideM + iter_key_start,
+          {problem_size_0_m, problem_size_0_n},
+          thread_id());
+        cutlass::TensorRef<scalar_t, cutlass::layout::RowMajor> bias_tensor_ref(
+          shared_storage.after_mm0.bias.data(),
+          cutlass::layout::RowMajor(MM0::ThreadblockShape::kN));
+        typename MM0::BiasLoader::SmemTileIterator smem_tile_iter(
+          bias_tensor_ref,
+          thread_id());
+        MM0::BiasLoader::load(bias_iter, smem_tile_iter);
+
+        // Pij += Bij, Pij is in register fragment and Bij is in shared memory
+        auto lane_offset = MM0::ScalingCoefsUpdater::get_lane_offset(
+            lane_id(), warp_id(), iteratorC_tile_offset);
+        MM0::ScalingCoefsUpdater::iterateRows(
+            lane_offset,
+            [&](int accum_m) {},
+            [&](int accum_m, int accum_n, int idx) {
+              if (accum_m < problem_size_0_m && accum_n < problem_size_0_n) {
+                accum[idx] += bias_tensor_ref.at({accum_m, accum_n});
+              }
+            },
+            [&](int accum_m) {});
+      }
+
       // Mask out last if causal
       if (p.causal && p.num_keys - iter_key_start <= kKeysPerBlock) {
```

Comment:
Interesting - we can still have causal masking + custom mask. We would need to find a way to expose that properly in a follow-up PR

Reply:
yep. i think that for users that need both additive bias and unidirectionality, using causal masking in addition to attention bias might be faster than just using attention bias with some of the values replaced with -inf since the kernel can use the knowledge that its causal to avoid some unnecessary compute. if thats true it might be unintuitive to some users though